tweak: redundancy checks for control input

Spudgun opened the pull request:
Enables key binds for poly manipulation regardless if a key is enabled or disabled.

A lot of no-clips disable input of all types, making the IsControlJustPressed not trigger. Essentially just makes this script compatible with a lot of other scripts that disable input of certain types.

Ekinoxx0 commented:
From my testing IsDisabledControlPressed also return valid data even if the control is not disabled, it seems to me that IsControlKeyJustPressed is redundant from just using the Disable native.

mkafrin commented:
Hmm, @Spudgun can you check if what Ekinoxx said is accurate, and if so, update the PR?

Spudgun commented:
> Hmm, @Spudgun can you check if what Ekinoxx said is accurate, and if so, update the PR?
> 
> Also, the difference between "ControlPressed" and "ControlJustPressed" in the resource is intentional. In some cases, things should only work on press and others (like modifiers) should support pressing and holding. Combining those two is fundamentally changing how the controls work and I don't think that's good. If you had a particular reason for doing this, please detail that reason so I can better understand.

So I've just had a play around with the control enabling/disabling. I can confirm that what @Ekinoxx0 said is correct. If a control key is enabled or disabled right before the check that of `IsDisabledControlPressed` or `IsControlPressed` etc will then trigger even if the control is being disabled/enabled in another script.

This is then making my implementation of `IsControlKeyJustPressed` function redundant.

In saying all of the above, I have updated my PR to ensure that functionality is still possible in the circumstances another script is disabling the control we're testing for. I have also reverted the `JustPressed` and `Press` changes as I didn't think through the intended use case scenario for those. It's quite fundamental to how the script works, as you said 🙂

mkafrin commented on creation/client/BoxZone.lua:
```diff
@@ -3,12 +3,17 @@ local function handleInput(useZ, heading, length, width, center)
-    DisableControlAction(0, 81, true)
-    if IsDisabledControlJustPressed(0, 81) then
+    if IsDisabledControlJustPressed(0, 81) then -- scroll wheel down just pressed
+      EnableControlAction(0, 19, true)
+      EnableControlAction(0, 21, true)
+
       if IsControlPressed(0, 19) then -- alt held down
```
So it seems like Ekin is correct that `IsDisabledControlPressed` works even when the control is not disabled. Do you have anything against removing these `EnableControlAction` calls and just using `IsDisabledControlPressed` and `IsDisabledControlJustPressed` in all places (for this, for pressing Z on line 101, etc.)? I don't want to errantly re-enable controls if other resources have chosen to disable them.

Spudgun replied:
The only issue I see with this is if the control isn't actually disabled the check won't go through. As it's checking for a disabled control press. Meaning that we need to enable/disable the control for the test case of input.

I think it is safe to change the control input as I assume if the user is setting up a PolyZone that they're actively using this script and not others. Happy to go with whatever you'd want though 😊

mkafrin replied:
> The only issue I see with this is if the control isn't actually disabled the check won't go through. As it's checking for a disabled control press. Meaning that we need to enable/disable the control for the test case of input.

So that's actually what I'm saying. I thought this was the case but I just tested it and it seems even if the control is enabled the check will still go through. It seemed even if I put `EnableControlAction` right before it and then called `IsDisabledControlPressed` it still passed the check. But now that I think on it some more, that feels more like a bug and prob not good to rely on it always being the case.

Instead, let's just go all in on disabling. Do you have anything against flipping all the `EnableControlAction` to `DisableControlAction` and `IsControlPressed`/`IsControlJustPressed` to `IsDisabledControlPressed`/`IsDisabledControlJustPressed`? I don't think that'll break anything and just disabling everything makes sense to me.

Spudgun replied:
That's completely fine. 

I just tested around with DisabledControl for everything and it still works, which as you said must be a bug? I will be interested to try this on a vanilla server and see if the thing happens. Seems weird to me that `IsDisabledControlPressed`/`IsDisabledControlJustPressed` trigger even if the control isn't disabled.

In regards to changing everything, I can do that. However, in case this is a bug it might be worthwhile disabling those controls in a thread whenever a user uses `/pzcreate`. What are your thoughts? That way if it does get patched it won't break the script. See below for the code example.

**creation.lua**
```lua
DisableControlKeyInput = function()
  if not drawZone then
    Citizen.CreateThread(function()
      while drawZone do
        Citizen.Wait(0)
        DisableControlAction(0, 36, true)   -- Ctrl
        DisableControlAction(0, 19, true)   -- Alt
        DisableControlAction(0, 20, true)   -- 'Z'
        DisableControlAction(0, 21, true)   -- Shift
        DisableControlAction(0, 81, true)   -- Scroll Wheel Down
        DisableControlAction(0, 99, true)   -- Scroll Wheel Up

        DisableControlAction(0, 172, true)  -- Arrow Up
        DisableControlAction(0, 173, true)  -- Arrow Down
        DisableControlAction(0, 174, true)  -- Arrow Left
        DisableControlAction(0, 175, true)  -- Arrow Right

      end
    end)
  end
end
```

**It would get called here..**
```lua
RegisterNetEvent("polyzone:pzcreate")
AddEventHandler("polyzone:pzcreate", function(zoneType, name, args)
  if createdZone ~= nil then
    TriggerEvent('chat:addMessage', {
      color = { 255, 0, 0},
      multiline = true,
      args = {"Me", "A shape is already being created!"}
    })
    return
  end

  if zoneType == 'poly' then
    polyStart(name)
  elseif zoneType == "circle" then
    local radius = nil
    if #args >= 3 then radius = tonumber(args[3])
    else radius = tonumber(GetUserInput("Enter radius:")) end
    if radius == nil then
      TriggerEvent('chat:addMessage', {
        color = { 255, 0, 0},
        multiline = true,
        args = {"Me", "CircleZone requires a radius (must be a number)!"}
      })
      return
    end
    circleStart(name, radius)
  elseif zoneType == "box" then
    local length = nil
    if #args >= 3 then length = tonumber(args[3])
    else length = tonumber(GetUserInput("Enter length:")) end
    if length == nil or length < 0.0 then
      TriggerEvent('chat:addMessage', {
        color = { 255, 0, 0},
        multiline = true,
        args = {"Me", "BoxZone requires a length (must be a positive number)!"}
      })
      return
    end
    local width = nil
    if #args >= 4 then width = tonumber(args[4])
    else width = tonumber(GetUserInput("Enter width:")) end
    if width == nil or width < 0.0 then
      TriggerEvent('chat:addMessage', {
        color = { 255, 0, 0},
        multiline = true,
        args = {"Me", "BoxZone requires a width (must be a positive number)!"}
      })
      return
    end
    boxStart(name, 0, length, width)
  else
    return
  end
  createdZoneType = zoneType
  drawZone = true
  DisableControlKeyInput()
  drawThread()
end)
```

**And here...**
```lua
RegisterNetEvent("polyzone:pzlast")
AddEventHandler("polyzone:pzlast", function()
  if createdZone ~= nil or lastCreatedZone == nil then
    return
  end
  if lastCreatedZoneType == 'poly' then
    TriggerEvent('chat:addMessage', {
      color = { 0, 255, 0},
      multiline = true,
      args = {"Me", "The command pzlast only supports BoxZone and CircleZone for now"}
    })
  end

  local name = GetUserInput("Enter name (or leave empty to reuse last zone's name):")
  if name == nil then
    return
  elseif name == "" then
    name = lastCreatedZone.name
  end
  createdZoneType = lastCreatedZoneType
  if createdZoneType == 'box' then
    local minHeight, maxHeight
    if lastCreatedZone.minZ then
      minHeight = lastCreatedZone.center.z - lastCreatedZone.minZ
    end
    if lastCreatedZone.maxZ then
      maxHeight = lastCreatedZone.maxZ - lastCreatedZone.center.z
    end
    boxStart(name, lastCreatedZone.offsetRot, lastCreatedZone.length, lastCreatedZone.width, minHeight, maxHeight)
  elseif createdZoneType == 'circle' then
    circleStart(name, lastCreatedZone.radius, lastCreatedZone.useZ)
  end
  drawZone = true
  DisableControlKeyInput()
  drawThread()
end)
```

Let me know your thoughts.

mkafrin replied:
Yeah I have no problem with that. It's cleaner anyways.

To avoid a separate comment later, can you also put the `DisableControlKeyInput` in `utils.lua`, create the function using the same syntax we use elsewhere `function DisableControlKeyInput()`, remove the `Citizen.` from before `Wait`, and put the `Wait` at the bottom of the while loop, to follow what we do elsewhere?

Spudgun replied:
All done 😊

mkafrin commented on creation/client/creation.lua:
```diff
@@ -59,6 +59,7 @@ AddEventHandler("polyzone:pzcreate", function(zoneType, name, args)
   end
   createdZoneType = zoneType
   drawZone = true
+  disableControlKeyInput()
```
One last thing. This is called after drawZone is set to true, but then the function does nothing if drawZone is true. So the controls are not actually being disabled. This call should happen before the previous line. Same for below (line 127) as well.

Spudgun replied:
Whoops. Have updated the PR to reflect :)

mkafrin replied:
Works for me! Thank you.

mkafrin reviewed (approved):
Thank you for bringing this up and contributing to fix it. And I appreciate you engaging in the conversation around the best way to make the changes.

Spudgun commented:
All good, mate! 

Happy to help contribute to such a useful tool!
